hnswlib/hnswlib.h

```diff
@@ -4,6 +4,8 @@
 #define USE_SSE
 #ifdef __AVX__
 #define USE_AVX
+#ifdef __AVX512__
```

Comment:
Hm. It looks like this `#ifdef` does not have a corresponding `#endif` and this seems to trigger the CI build fail for some systems.

Reply:
Yes, I did test with multiple concurrent processes running. We plugged the functions:  InnerProductSIMD16Ext and  L2SqrSIMD16Ext into a different application. The new application was implementing a brute-force vector similarity search.
The test was performed on c5.4xlarge on AWS. 

We can see that AVX512 is consistently better (lower execution time) than AVX2. We can also see that after about 8 parralel instances, the latency shoots up dramatically. We believe that this is due to memory latency. The brute-force algorithm is a memory scan and thus very memory bandwith intensive. After about 8 concurrent instances, I expect that we are saturating the memory bandwidth of the CPU. 

![IP_brute_force](https://user-images.githubusercontent.com/8259383/132893128-e55c842a-51b1-4921-a8ea-1d52ce3e0110.png)
![L2_brute_force](https://user-images.githubusercontent.com/8259383/132893131-d38f169d-5ca6-4910-a2ad-9ce381f88d15.png)
 
It is true that frequency goes down with AVX512 as compared with AVX2. Same is also true for AVX2 vs AVX. The number of cores executing AVX/AVX2/AVX512 code, determines the frequency. The 3 figures below, show the frequency tables depending on how many cores are active. As an example, I have circled a single point - assuming all cores active on a 8280 CPU (which is CascadeLake - same generation as used in c5 AWS instances). 

We can see that with all cores active, AVX code will run at 3.3GHz, AVX2 code will run at 2.9GHz, and AVX512 code will run at 2.4GHz. Even though, there is a frequency drop, we are operating on 2X the data at a time. So, if we spend enough time in the computation, then AVX512 will be a net benefit. In addition, AVX512 frequency is improving in subsequent generations - e.g. The m6i instance (IceLake ) is bridging the gap between AVX2 and AVX512 frequency. 

![AVX_frequency](https://user-images.githubusercontent.com/8259383/132895590-957b01f5-9cee-42ad-8bbe-bfb9b2bc590e.png)
![AVX2_frequency](https://user-images.githubusercontent.com/8259383/132894388-2500680c-3ab6-40be-a2bf-f2ad71cdd3d3.png)
![AVX512_frequency](https://user-images.githubusercontent.com/8259383/132894391-46e401b2-841c-4a81-87d3-8a8e3ac20ff2.png)

Reply:
> Hm. It looks like this `#ifdef` does not have a corresponding `#endif` and this seems to trigger the CI build fail for some systems.

Thank you for catching this. I have added the missing `#endif`

Reply:
The initial ann-bechmark results that I posted were with parallelism=3 on an c5.4xlarge instance. 
I just tested with parallelism=6 and the results are similar. I tried with parallelism=8, but we run out of memory on c5.4xlarge. 

![image](https://user-images.githubusercontent.com/8259383/132992341-38ac14ef-b014-497c-825f-bd7d3f8dfb9c.png)
![image](https://user-images.githubusercontent.com/8259383/132992350-df1885c8-9c29-4a54-805f-693875b34217.png)
![image](https://user-images.githubusercontent.com/8259383/132992356-c52d835a-fc40-4115-ac9c-76a0c24e7288.png)
![image](https://user-images.githubusercontent.com/8259383/132992359-442c0143-fc4a-456c-872e-511507ca9847.png)

Reply:
Thanks for the test and explanation!